### PR TITLE
Filtered Lifting Line Theory Correction for the Actuator Line Model

### PR DESCRIPTION
**Pull request by tonyinme**
**Pull-request type:**
- New Feature

## Description of the pull-request

This pull request adds the new filtered lifting line theory correction to the actuator line model.
It allows the use of coarse resolutions/epsilon and the velocities are corrected as if a smaller epsilon value is used.

A new unit test is added for the Gaussian function used in the actuator section.

## Checklist

*All pull requests*
- Builds successfully
  - Operating systems 
    -  Linux
    -  MacOS
  - Compilers 
    -  LLVM/Clang
    -  Intel compilers
-  Compiles without warnings

### CODE REVIEW

File: src/actuator/ActuatorFAST.C [hunk 1]
```diff
@@ -682,6 +735,361 @@ ActuatorFAST::execute()
...
+          auto infoObject =
+            dynamic_cast<ActuatorFASTPointInfo*>(
+              actuatorPointInfoMap_.at(np).get());
+        
+          // Get the force from FAST
+          FAST.getForce(force, np, infoObject->globTurbId_);
+          // Get the velocity from FAST
+          FAST.getRelativeVelForceNode(vel, np, infoObject->globTurbId_);
+
+          // Zero the radial component of the velocity and force
```
Comment (psakievich) [inline src/actuator/ActuatorFAST.C]: Not critical, but removing some of these stale comments from debugging would be nice.

**Comment by sayerhs** [comment]
@tonyinme As @psakievich mentioned, we need to remove all auto-generated files from the pull-request. The `CMakeFiles` directory should be removed. An additional thing I notice is that this pull request creates more duplicates of NREL 5MW OpenFAST input deck. We had discussed this #183 without a clear resolution, and that is a bit of a concern.

File: examples/filtered_lifting_line/with_correction/5MW_Baseline/Airfoils/Cylinder1.dat [hunk 2]
```diff
@@ -0,0 +1,58 @@
+! ------------ AirfoilInfo v1.01.x Input File ----------------------------------
```
Comment (sayerhs) [inline examples/filtered_lifting_line/with_correction/5MW_Baseline/Airfoils/Cylinder1.dat]: Please remove duplicate 5MW input files. Use `symlink` and refer to the input deck in regression tests.

File: include/actuator/UtilitiesActuator.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/*------------------------------------------------------------------------*/
```
Comment (sayerhs) [inline include/actuator/UtilitiesActuator.h]: Update license to new joint-copyright boilerplate

File: src/actuator/ActuatorFAST.C [hunk 4]
```diff
@@ -43,13 +43,21 @@
 #include <stdexcept>
 #include <cmath>
 
+// The utilities used for actuator
+#include "actuator/UtilitiesActuator.h"
+
+const double pi = 3.14159265358979323846;
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Enclose this in an private namespace

File: src/actuator/ActuatorFAST.C [hunk 5]
```diff
@@ -560,6 +597,8 @@ ActuatorFAST::execute()
   // loop over map and get velocity at points
   for (std::size_t np = 0; np < numFastPoints_; np++) {
 
+//~ std::cerr << "Execute before ==============================" << np << " " << NaluEnv::self().parallel_rank() << std::endl;
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Remove

File: src/actuator/ActuatorFAST.C [hunk 6]
```diff
@@ -610,9 +649,23 @@ ActuatorFAST::execute()
       1, bestElem, bulkData, infoObject->isoParCoords_.data(), &ws_density_[0],
       &ws_pointGasDensity);
     int nNp = (int)np;
+    
+    /////////////////////////
+    // Add the filtered lifting line theory correction here
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Is this comment valid?

File: src/actuator/ActuatorFAST.C [hunk 7]
```diff
@@ -682,6 +735,361 @@ ActuatorFAST::execute()
   stk::mesh::parallel_sum_including_ghosts(bulkData, sumFieldG);
 }
 
+
+// Copmute the filtered lifting line theory correction
+void ActuatorFAST::filtered_lifting_line()
+{
+
+
+  // The number of dimensions (assumes 3D)
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Remove unnecessary comment

File: src/actuator/ActuatorFAST.C [hunk 8]
```diff
@@ -682,6 +735,361 @@ ActuatorFAST::execute()
...
+  int nDim=3;
+
+  // The total number of turbines
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Remove unnecessary comment

File: src/actuator/ActuatorFAST.C [hunk 9]
```diff
@@ -682,6 +735,361 @@ ActuatorFAST::execute()
...
+  // points in the simulation
+  size_t np;
+
+
+  // Loop through all turbines
+  for (size_t iTurb=0; iTurb < numTurbines; iTurb++ )
+  {
+
+    // Only use if correction is active for this turbine
+    // Point to the last element of this array that was just appended
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Wrong comment for the code in question

File: src/actuator/ActuatorFAST.C [hunk 10]
```diff
@@ -682,6 +735,361 @@ ActuatorFAST::execute()
...
+  // The total number of turbines
+  const size_t numTurbines = fi.nTurbinesGlob;
+
+  // The index of actuator point
+  // This identifiex an actuator point in a vector containing all actuator
+  // points in the simulation
+  size_t np;
+
+
+  // Loop through all turbines
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Remove redundant comment

File: src/actuator/ActuatorFAST.C [hunk 11]
```diff
@@ -682,6 +735,361 @@ ActuatorFAST::execute()
...
+            // Get the adjacent index
+            np = indexMap_[iTurb][nb][na + 1];
+            // Get the coordinates for the current point
+            FAST.getForceNodeCoordinates(xyz_p1, np, iTurb);
+
+            // Compute the magnitude of the difference
+            for (int i = 0; i < nDim; i++)
+                dr += std::pow(xyz_p1.data()[i]-xyz.data()[i], 2);
+
+            // Take the square root and divide by 2 (central difference)
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: This is redundant

File: src/actuator/ActuatorFAST.C [hunk 12]
```diff
@@ -932,13 +1356,221 @@ ActuatorFAST::create_actuator_point_info_map()
...
+  int na = 0;
+  // Loop through all turbines and only call FAST functions if processor owns 
+  //   the turbine
+  for (int i = 0; i < nt; ++i) {
+    if (FAST.get_procNo(i) != NaluEnv::self().parallel_rank()) continue;
+    nb = std::max(nb, FAST.get_numBlades(i));
+    na = std::max(na, fi.globTurbineData[i].numForcePtsBlade);
+  }
+
+// BROKEN CODE
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Remove

File: src/actuator/ActuatorFAST.C [hunk 13]
```diff
@@ -989,10 +1622,18 @@ ActuatorFAST::spread_actuator_force_to_node_vec(
   std::set<stk::mesh::Entity>::iterator iNode;
   for (iNode = nodeVec.begin(); iNode != nodeVec.end(); ++iNode) {
 
+//~ std::cerr << "spread in before =======" << *iNode << " " << NaluEnv::self().parallel_rank() << std::endl;
+
     stk::mesh::Entity node = *iNode;
 
+////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Has this been fixed?

File: src/actuator/UtilitiesActuator.C [hunk 14]
```diff
@@ -0,0 +1,74 @@
+/*------------------------------------------------------------------------*/
```
Comment (sayerhs) [inline src/actuator/UtilitiesActuator.C]: Update copyright to new joint-copyright boilerplate

File: unit_tests/UnitTestActuatorFASTGaussian.C [hunk 15]
```diff
@@ -0,0 +1,63 @@
+/*------------------------------------------------------------------------*/
+/*  Copyright 2014 Sandia Corporation.                                    */
```
Comment (sayerhs) [inline unit_tests/UnitTestActuatorFASTGaussian.C]: Update copyright notice

File: unit_tests/UnitTestActuatorFASTGaussian.C [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+/*  in the file, LICENSE, which is located in the top-level Nalu          */
+/*  directory structure                                                   */
+/*------------------------------------------------------------------------*/
+/*
+ * UnitTestActuatorDiskFAST.C
+ *
+ *  Created on: Nov, 2019
+ *      Author: tony
+ */
+
```
Comment (sayerhs) [inline unit_tests/UnitTestActuatorFASTGaussian.C]: This information is available in git history and `git-blame`, we haven't been putting individual names in files.

File: examples/filtered_lifting_line/with_correction/alm_simulation.yaml [hunk 17]
```diff
@@ -0,0 +1,176 @@
+Simulations:
```
Comment (psakievich) [inline examples/filtered_lifting_line/with_correction/alm_simulation.yaml]: Both of these example runs are segfaulting in OpenFAST when I test them locally.  Not sure why. Something about and index of 1 being greater than the max accepted value of 0.  The regression test worked when I tested it.

Comment (sayerhs) [reply examples/filtered_lifting_line/with_correction/alm_simulation.yaml]: @psakievich I see that you approved this pull-request. Has this been fixed?

File: include/actuator/UtilitiesActuator.h [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+// for more details.
+
+namespace sierra {
+namespace nalu {
+
+struct Coordinates;
+
+namespace actuator_utils {
+
+// A Gaussian projection function
```
Comment (sayerhs) [inline include/actuator/UtilitiesActuator.h]: Would be nice if these were doxygen compatible docstrings.

File: src/actuator/ActuatorFAST.C [hunk 19]
```diff
@@ -43,13 +43,19 @@
...
 namespace sierra {
 namespace nalu {
 
 // constructor
 ActuatorFASTInfo::ActuatorFASTInfo() : ActuatorInfo()
 {
-  // nothing to do
+  // Initialize the filtered lifting line theory correciton to false
+  // This is later read from the yaml input file and can be changed
+  fllt_correction_ = false;
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Better to do this as either in the header or in the c'tor initializer list

File: src/actuator/ActuatorFAST.C [hunk 20]
```diff
@@ -65,12 +71,15 @@ ActuatorFASTPointInfo::ActuatorFASTPointInfo(
   double searchRadius,
   // The values of epsilon [m]
   Coordinates epsilon,
+  // The optimal epsilon [m]
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Why not use doxygen-compatible docstrings here?

File: src/actuator/ActuatorFAST.C [hunk 21]
```diff
@@ -201,7 +178,10 @@ ActuatorFAST::load(const YAML::Node& y_node)
           const YAML::Node cur_turbine =
             y_actuator["Turbine" + std::to_string(iTurb)];
 
+          // Append a new object to the arrary
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Avoid useless comments like these.

File: src/actuator/ActuatorFAST.C [hunk 22]
```diff
@@ -201,7 +178,10 @@ ActuatorFAST::load(const YAML::Node& y_node)
...
           actuatorInfo_.emplace_back(new ActuatorFASTInfo());
+
+          // Point to the last element of this array that was just appended
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Avoid useless comments

File: src/actuator/ActuatorFAST.C [hunk 23]
```diff
@@ -214,6 +194,18 @@ ActuatorFAST::load(const YAML::Node& y_node)
             actuatorFASTInfo->fileToDumpPoints_ = turbFileName;
           }
 
+          // The correction from filtered lifting line theory
+          //~ const bool fllt_correction = 
+                  //~ cur_turbine["fllt_correction"].as<bool>();
+          bool fllt_correction=false;
+          get_if_present(cur_turbine, "fllt_correction", fllt_correction);
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: You most likely want `get_if_present_no_default` here, or use 

```C++
get_if_present(cur_turbine, "fllt_correction", fllt_correction, fllt_correction);
```

File: src/actuator/ActuatorFAST.C [hunk 24]
```diff
@@ -214,6 +194,18 @@ ActuatorFAST::load(const YAML::Node& y_node)
...
+          if (fllt_correction) {
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Redundant code.

File: src/actuator/ActuatorFAST.C [hunk 25]
```diff
@@ -932,13 +1335,189 @@ ActuatorFAST::create_actuator_point_info_map()
...
+        NaluEnv::self().naluOutput()
+            << "Proc " << NaluEnv::self().parallel_rank() << " glob iTurb "
+            << iTurb << std::endl;
+      }
+    }
+  }
+   create_point_info_map_class_specific(); 
+}
+
+// This function computes the index map such that actuator points can be
```
Comment (sayerhs) [inline src/actuator/ActuatorFAST.C]: Doxygen compatible comments here?

File: src/actuator/UtilitiesActuator.C [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+
+
+namespace sierra{
+namespace nalu {
+namespace actuator_utils {
+
+// This is the variable pi
+const double pi = M_PI;
+
+// A Gaussian projection function
```
Comment (sayerhs) [inline src/actuator/UtilitiesActuator.C]: Use doxygen compatible commenting style

**Comment by sayerhs** [comment]
@psakievich Can you move one of the `examples` into a regression test so that the FLLT option is exercised for the nightly tests on all compilers? It is fine if it is retained in examples also, but I'd like to see that we catch regressions/issues on a wide range of compilers and make sure that option is always exercised.

**Comment by psakievich** [comment]
@sayerhs sure thing.

File: examples/filtered_lifting_line/with_correction/alm_simulation.yaml (hunk 17 above)
Comment (psakievich) [reply examples/filtered_lifting_line/with_correction/alm_simulation.yaml]: fixed late last week

File: src/actuator/ActuatorFAST.C (hunk 25 above)
Comment (psakievich) [reply src/actuator/ActuatorFAST.C]: Okay

File: src/actuator/ActuatorFAST.C (hunk 23 above)
Comment (psakievich) [reply src/actuator/ActuatorFAST.C]: done

**Comment by psakievich** [comment]
@sayerhs I've put a PR in with @tonyinme and included updating some comments to be doxygen compatible.  I'm a little rusty with doxygen, so not sure if I did everything correctly, but as long as the code is good I'd like to merge this.  I've started refactoring on a local branch and can address a lot of the other details while I comb through this section of the code.

File: src/actuator/ActuatorFAST.C (hunk 19 above)
Comment (psakievich) [reply src/actuator/ActuatorFAST.C]: good catch

**Pull request merged**